### PR TITLE
Contracts remove deposit accounts

### PR DESCRIPTION
**Pull request by juangirini**
## In this PR we are removing the deposit accounts.

This is a follow up of https://github.com/paritytech/substrate/pull/14020 in which we got rid of the Currency traits in favour of the fungible traits, and therefore we *hold* deposits instead of *reserving* them.

ED doesn't contribute to *held* balances and this balance can only be slashed by the same `HoldReason` it was created under, removing the need of maintaining the deposit accounts.

## Migrations
Migrations v10 and v13 operated over deposit accounts and they have been updated to be kept backwards compatible for those chain which haven't ran it yet.

A new migration v15 has been added that moved the balance on the deposit accounts back to the contract accounts and hold them there under the `StorageDepositReserve` reason.

This new migration needs to be added to the Config
```rust
impl Config for Runtime {
	// ...
	type Migrations = (
		// ...
		v15::Migration<Runtime>,
	);
	// ...
}
```
---
Cumulus companion: https://github.com/paritytech/cumulus/pull/2973
Complementary PR https://github.com/paritytech/substrate/pull/14767

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3203651

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3209550

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3210418

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3221735

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3221733

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3224281

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3228301

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3234004

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3248835

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3258677

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3259211

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3281927

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3282043

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3300167

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3311012

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3319073

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3319074

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3320118

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3320132

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3320527

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3324126

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3324140

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3325057

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3325368

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3325353

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3326398

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3326384

**Comment by juangirini** [comment]
bot bench $ pallet dev pallet_contracts

**Comment by juangirini** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

**Comment by juangirini** [comment]
bot bench $ pallet dev pallet_contracts

**Comment by juangirini** [comment]
bot bench substrate-pallet --pallet=pallet_contracts

File: frame/contracts/src/storage/meter.rs [hunk 1]
```diff
@@ -533,31 +527,70 @@ impl<T: Config> Ext<T> for ReservingExt {
 
 	fn charge(
 		origin: &T::AccountId,
-		deposit_account: &DepositAccount<T>,
+		contract: &T::AccountId,
 		amount: &DepositOf<T>,
-		terminated: bool,
+		_terminated: bool,
```
Comment (pgherveou) [inline frame/contracts/src/storage/meter.rs:532]: let's remove if not needed anymore

File: frame/contracts/src/storage/meter.rs [hunk 2]
```diff
@@ -533,31 +527,70 @@ impl<T: Config> Ext<T> for ReservingExt {
...
+					// This should never happen, if it does it means that there is a bug in the
+					// runtime logic. In the rare case this happens we try to refund as much as we
+					// can, thus the `Precision::BestEffort`.
+					log::error!(
+						target: "runtime::contracts",
+						"Failed to repatriate full storage deposit {:?} from contract {:?} to origin {:?}. Transferred {:?}.",
+						amount, contract, origin, transferred,
+					);
+				} else {
+					let _ = System::<T>::dec_providers(contract);
```
Comment (pgherveou) [inline frame/contracts/src/storage/meter.rs:589]: why do we call dec_providers here?

**Review by pgherveou — COMMENTED** [review]
Looks good, just left a few comments / questions

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3334867

File: frame/contracts/src/migration/v15.rs [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+			))
+		} else {
+			old::ContractInfoOf::<T>::iter()
+		};
+
+		if let Some((account, old_contract)) = iter.next() {
+			let deposit_account = &old_contract.deposit_account;
+			if System::<T>::consumers(deposit_account) > Zero::zero() {
+				System::<T>::dec_consumers(deposit_account);
+			}
```
Comment (athei) [inline frame/contracts/src/migration/v15.rs:153]: We should remove the conditional. We **know** it is at least one. Also there is a check in this function which will print an error in this case instead of just silently doing nothing.

File: frame/contracts/src/migration/v15.rs [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+
+			// Hold the reserved balance.
+			if transferred_deposit_balance == Zero::zero() {
+				log::warn!(
+					target: LOG_TARGET,
+					"No balance to hold as storage deposit on the contract 0x{:?}.",
+					HexDisplay::from(&account.encode())
+				);
+			} else {
+				T::Currency::hold(
```
Comment (athei) [inline frame/contracts/src/migration/v15.rs:215]: Can't we just use `transfer_and_hold`?

**Review by athei — CHANGES_REQUESTED** [review]

File: frame/contracts/src/migration/v10.rs [hunk 5]
```diff
@@ -162,8 +174,7 @@ where
 			log::debug!(target: LOG_TARGET, "Account: 0x{} ", HexDisplay::from(&account.encode()));
 
 			// Get the new deposit account address
-			let deposit_account: DepositAccount<T> =
-				DepositAccount(T::AddressGenerator::deposit_address(&account));
+			let deposit_account: DepositAccount<T> = DepositAccount(deposit_address::<T>(&account));
```
[the remark this answers is not in the document]
Comment (juangirini) [reply frame/contracts/src/migration/v10.rs:177]: I would mark this as resolved if you agree, as mentioned by Alex [here](https://github.com/paritytech/substrate/pull/14589#discussion_r1285255665)

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3341023

File: frame/contracts/src/migration/v15.rs (hunk 4 above)
Comment (juangirini) [reply frame/contracts/src/migration/v15.rs:215]: This approach gives us a more granular control for better logging on batch operations like this. `transfer_and_hold` could very well fail at `decrease_balance` or `increase_balance_on_hold` and we won't know which one without inspecting the error.
Anyway, I am happy to use `transfer_and_hold` if that makes the code more readable.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3341888

File: frame/contracts/src/storage/meter.rs [hunk 6]
```diff
@@ -94,6 +98,16 @@ pub trait Ext<T: Config> {
...
+// /// Indicates the reason that originated the charge.
+// pub enum ChargeReason {
+// 	/// The charge was originated by a contract instantiation.
+// 	Instantiate,
+// 	/// The charge was originated by a contract termination.
+// 	Terminate,
+// 	/// The charge was not originated by a contract instantiation or termination.
+// 	Regular,
+// }
+
```
[the remark this answers is not in the document]
Comment (juangirini) [reply frame/contracts/src/storage/meter.rs]: This was old code while the PR was in draft mode. Removed

File: frame/contracts/src/storage/meter.rs (hunk 1 above)
Comment (juangirini) [reply frame/contracts/src/storage/meter.rs]: it has been brought back

File: frame/contracts/src/storage/meter.rs (hunk 2 above)
Comment (juangirini) [reply frame/contracts/src/storage/meter.rs]: During the termination process we first transfer all the free balance of the contract to the beneficiary:
https://github.com/paritytech/substrate/blob/810af1c564b3b38ea91838812203751c729599e7/frame/contracts/src/exec.rs#L1277-L1282

And then we release the held balance under the `StorageDepositReserve` reason and transfer it back to the beneficiary:
https://github.com/paritytech/substrate/blob/810af1c564b3b38ea91838812203751c729599e7/frame/contracts/src/storage/meter.rs#L562-L570

If we had no provider, the former `transfer()` would not include the `ed` as it needs it to keep the account alive (even while having `Preservation::Expendable` - see [`reducible_balance`](https://github.com/paritytech/substrate/blob/810af1c564b3b38ea91838812203751c729599e7/frame/balances/src/impl_fungible.rs#L44)), and therefore we'd need to deal with this `ed` somehow later on. 
The best solution I see then is to have a provider. 

Having said that, your concerns regarding the unbalanced `inc_providers`/`dec_providers` are correct and therefore I have moved them to pretty much replace where the previous `inc_consumers`/`dec_consumers` were, to make sure we only have one of each.

File: frame/contracts/src/storage/meter.rs [hunk 7]
```diff
@@ -533,31 +530,69 @@ impl<T: Config> Ext<T> for ReservingExt {
...
+					// runtime logic. In the rare case this happens we try to refund as much as we
+					// can, thus the `Precision::BestEffort`.
+					log::error!(
+						target: LOG_TARGET,
+						"Failed to repatriate full storage deposit {:?} from contract {:?} to origin {:?}. Transferred {:?}.",
+						amount, contract, origin, transferred,
+					);
+				} else if terminated {
+					System::<T>::dec_providers(&contract)?;
+				}
```
Comment (juangirini) [inline frame/contracts/src/storage/meter.rs:592]: I'd never reach this point if `amount.is_zero()`, but termination always carries at least the contract's info storage deposit so we should be good. Still, I'd like some feedback on whether you consider to be safer to check for termination even when `amount` is zero.

File: frame/contracts/src/migration/v13.rs [hunk 8]
```diff
@@ -20,25 +20,24 @@
...
 	migration::{IsFinished, MigrationStep},
-	storage::DepositAccount,
 	weights::WeightInfo,
-	BalanceOf, CodeHash, Config, Pallet, TrieId, Weight, LOG_TARGET,
+	AccountIdOf, BalanceOf, CodeHash, Config, Pallet, TrieId, Weight, LOG_TARGET,
 };
 use codec::{Decode, Encode};
 use frame_support::{codec, pallet_prelude::*, storage_alias, DefaultNoBound};
+#[cfg(feature = "runtime-benchmarks")]
+use sp_runtime::traits::{Hash, TrailingZeroInput};
```
Comment (athei) [inline frame/contracts/src/migration/v13.rs:29]: Move into the function to save one conditional compilation attribute.

File: frame/contracts/src/migration/v15.rs [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+		fungible::{Mutate, MutateHold},
+		tokens::{fungible::Inspect, Fortitude, Preservation},
+	},
+	BoundedBTreeMap, DefaultNoBound,
+};
+use frame_system::Pallet as System;
+use sp_core::hexdisplay::HexDisplay;
+#[cfg(feature = "runtime-benchmarks")]
+use sp_runtime::traits::{Hash, TrailingZeroInput};
+#[cfg(feature = "try-runtime")]
```
Comment (athei) [inline frame/contracts/src/migration/v15.rs:45]: Let's move it into the function or use fully qualified names to reduce the amount of conditional compilation attributes.

File: frame/contracts/src/storage/meter.rs [hunk 10]
```diff
@@ -441,21 +444,15 @@ where
 		// contract execution does conclude and hence would lead to a double charge.
 		self.total_deposit = deposit.clone();
 
+		// We need to make sure that the contract's account itself exists.
+		T::Currency::transfer(origin, contract, ed, Preservation::Preserve)?;
+
+		System::<T>::inc_providers(&contract);
```
Comment (athei) [inline frame/contracts/src/storage/meter.rs:450]: Need a comment here why this is a provider and not a `consumer` or `sufficient`. It was a consumer before.

File: frame/contracts/src/storage/meter.rs [hunk 11]
```diff
@@ -441,21 +444,15 @@ where
 		// contract execution does conclude and hence would lead to a double charge.
 		self.total_deposit = deposit.clone();
 
+		// We need to make sure that the contract's account itself exists.
```
Comment (athei) [inline frame/contracts/src/storage/meter.rs:447]:
```suggestion
		// We need to make sure that the contract's account exists.
```

File: frame/contracts/src/storage/meter.rs (hunk 2 above)
Comment (athei) [reply frame/contracts/src/storage/meter.rs]: Good. This was about incrementing and decrementing for each and every storage deposit. Not that we shouldn't have a provider.

File: frame/contracts/src/storage/meter.rs (hunk 7 above)
Comment (athei) [reply frame/contracts/src/storage/meter.rs:592]: Yeah let's remove this conditional. If we terminate we always want to remove the provider. Unbalanced providers is pretty bad.

Comment (juangirini) [reply frame/contracts/src/storage/meter.rs]: We need a way to check if `amount` is zero at some point as `transfer_and_hold` and `transfer_on_hold` don't do it. Otherwise we'd be calling them unnecessarily and we'd be depositing zero-amount events.

I have kept this conditional where it is, but moved the `dec_providers` outside of the `match` statement so that it is always called on a termination process.

File: frame/contracts/src/storage/meter.rs (hunk 10 above)
Comment (juangirini) [reply frame/contracts/src/storage/meter.rs]: This has been moved to (and commented) within the `terminate` fn
https://github.com/paritytech/substrate/blob/8ed3b66d41dc8518351180c9420a0df52b34fd56/frame/contracts/src/exec.rs#L1278-L1284

**Comment by juangirini** [comment]
bot rebase

File: frame/contracts/src/migration/v15.rs (hunk 4 above)
Comment (athei) [reply frame/contracts/src/migration/v15.rs:203]: Makes sense. Let's keep it this way then.

File: frame/contracts/src/storage/meter.rs (hunk 10 above)
Comment (athei) [reply frame/contracts/src/storage/meter.rs]: Okay by doing this we are able to send away all the balance and keep the account briefly alive. However, we should still add a `consumer` at account creation and remove it on termination as we did before (the provider here is on top). Reason being is that the system will be to brittle if we don't. Our guarantee is that: As long as a contract exists its account does exist. Without a consumer it hinges on us on calling transfer with the correct parameters. Something that we can't enforce if a contract calls into the runtime. With the consumer a correct runtime cannot remove the account.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3386557

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3393073

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3396043

File: frame/contracts/src/storage/meter.rs [hunk 12]
```diff
@@ -412,12 +415,12 @@ where
...
 	/// deposit charge separately from the storage charge.
-	pub fn charge_deposit(&mut self, deposit_account: DepositAccount<T>, amount: DepositOf<T>) {
+	pub fn charge_deposit(&mut self, contract: T::AccountId, amount: DepositOf<T>) {
 		self.total_deposit = self.total_deposit.saturating_add(&amount);
-		self.charges.push(Charge { deposit_account, amount, terminated: false });
+		self.charges.push(Charge { contract, amount, terminated: false });
 	}
 
-	/// Charges from `origin` a storage deposit for contract instantiation.
+	/// Charge from `origin` a storage deposit plus the ed for contract instantiation.
```
Comment (pgherveou) [inline frame/contracts/src/storage/meter.rs:423]: We transfer ed then charge deposit minus ed, so this comment or the code seems wrong 🤔

File: frame/contracts/src/storage/meter.rs [hunk 13]
```diff
@@ -441,21 +444,15 @@ where
...
-			&deposit.saturating_sub(&Deposit::Charge(ed)),
-			false,
-		)?;
-
-		System::<T>::inc_consumers(contract_info.deposit_account())?;
-
-		// We also need to make sure that the contract's account itself exists.
-		T::Currency::transfer(origin, contract, ed, Preservation::Preserve)?;
-		System::<T>::inc_consumers(contract)?;
+		E::charge(origin, contract, &deposit.saturating_sub(&Deposit::Charge(ed)), false)?;
```
Comment (pgherveou) [inline frame/contracts/src/storage/meter.rs:455]: comment above seems off as well now, since you start by transferring ed

File: frame/contracts/src/storage/meter.rs (hunk 12 above)
Comment (juangirini) [reply frame/contracts/src/storage/meter.rs:423]: `charge` is in itself a `transfer`, so I would not distinguish from transfer and charge. In fact we are taking "storage deposit + ed" from the origin.
How can we rephrase this to make it more clear?

File: frame/contracts/src/storage/meter.rs (hunk 13 above)
Comment (juangirini) [reply frame/contracts/src/storage/meter.rs:455]: Yeah, I don't think we need any comment at all there, maybe just the fact that we are excluding the `ed` from the `deposit` because it was transferred in the previous step. Would you add anything else?

File: frame/contracts/src/storage/meter.rs (hunk 12 above)
Comment (pgherveou) [reply frame/contracts/src/storage/meter.rs:423]: > In fact we are taking "storage deposit + ed" from the origin.

are you?
you do 
```rust
T::Currency::transfer(origin, contract, ed, Preservation::Preserve)?;
//...
E::charge(origin, contract, &deposit.saturating_sub(&Deposit::Charge(ed)), false)?;
```
you take `ed` and then `deposit - ed`, am I overlooking something here?

also since we transfer ed directly, couldn't we then use `fn charge_deposit` to delay the charge until later?

File: frame/contracts/src/storage/meter.rs (hunk 13 above)
Comment (pgherveou) [reply frame/contracts/src/storage/meter.rs:455]: yep let's just remove it, also see my comment above

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3408477

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3408476

File: frame/contracts/src/storage/meter.rs (hunk 12 above)
Comment (juangirini) [reply frame/contracts/src/storage/meter.rs:433]: > you take `ed` and then `deposit - ed`, am I overlooking something here?

yes, you're right

> also since we transfer ed directly, couldn't we then use `fn charge_deposit` to delay the charge until later?

absolutely, good point

File: frame/contracts/src/storage/meter.rs (hunk 10 above)
Comment (juangirini) [reply frame/contracts/src/storage/meter.rs]: this has been resolved in https://github.com/paritytech/substrate/pull/14767

**Comment by juangirini** [comment]
bot fmt

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3408788

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3408789

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3408790

File: frame/contracts/src/storage/meter.rs [hunk 14]
```diff
@@ -335,14 +346,13 @@ where
 		self.total_deposit.available(&self.limit)
 	}
 
-	/// True if the contract is alive.
-	fn is_alive(&self) -> bool {
-		matches!(self.own_contribution, Contribution::Alive(_))
-	}
```
Comment (pgherveou) [inline frame/contracts/src/storage/meter.rs:337]: I would keep it around seems easier to read than in lining `matches!(self.contract_state(), ContractState::Alive)` in a few places

**Review by pgherveou — APPROVED** [review]

File: frame/contracts/src/storage/meter.rs (hunk 14 above)
Comment (juangirini) [reply frame/contracts/src/storage/meter.rs:337]: could be, given the fact that we have `ContractState::Alive` and `Contribution::Alive()` the inline seems more meaningful... but I don't have a strong opinion here 
@agryaznov wdyt?

**Review by agryaznov — APPROVED** [review]

**Review by athei — APPROVED** [review]

**Comment by juangirini** [comment]
bot merge force
